packages/metro-react-native-babel-preset/src/configs/main.js

```diff
@@ -20,6 +20,15 @@ function isTSXSource(fileName) {
   return !!fileName && fileName.endsWith('.tsx');
 }
 
+function notTypeScriptSource(fileName) {
+  return !isTypeScriptSource(fileName) && !isTSXSource(fileName);
+}
+
+function optionsForTransformTS(opts) {
+  // `allowDeclareFields` will be enabled by default in Babel 8
+  return {...opts, allowDeclareFields: true, allowNamespaces: true};
+}
```

Comment:
Nit: I would instead create a constant e.g.
```
const COMMON_TS_OPTIONS = {
  // `allowDeclareFields` will be enabled by default in Babel 8
  allowDeclareFields: true,
  allowNamespaces: true,
};
```

and spread it into the usages below.

Also, do you have a reference for why Babel is treating this as a breaking change? Could this affect Metro users who for some reason need to not use this syntax yet?

Reply:
"why Babel is treating this as a breaking change?" - I'm not sure what you're asking. Here is the PR that adds the change to the Flow plugin: https://github.com/babel/babel/pull/11178
I think the reason it's breaking is that the default for `allowDeclareFields` is false. I'm not sure why it defaulted to that.

"could this affect Metro users...?" - I don't think it will. I tried to keep this low impact by (1) keeping the flow plugin in use for all files that are not `.ts[x]`, but switching TypeScript files to use the TypeScript transform instead of the Flow transform will cause some behavior changes. I hope that these changes are more desirable since using a Flow parser on TypeScript files could cause other unforeseen problems.

Reply:
> I think the reason it's breaking is that the default for `allowDeclareFields` is false

I guess my question is the other way around: presumably, the default for `allowDeclareFields` is false because the Babel team must have judged it to be a breaking change. So is there a reason for us to move separately from Babel on this? IMO, we should make it possible for people to configure Babel explicitly [per the Babel recommentation](https://babeljs.io/blog/2019/11/05/7.7.0#:~:text=it%20is%20recommended%20that%20you%20migrate%20your%20config%20to%20use%20it), and only switch to Babel 8's defaults when Babel 8 actually ships.

I fully agree that we should not be trying to run a Flow plugin on TypeScript code - that part is purely a bug.

Reply:
I believe I've found the breaking change thanks to your link to that PR. I found this PR for the original TypeScript feature: https://github.com/babel/babel/pull/10546
It would appear that `allowDeclareFields` affects the JS emitted for uninitialized fields. I was able to verify using the metro unit tests that given TypeScript:
```
class B {
  field1: string;
}
```
The code emitted in the class constructor with `allowDeclareFields: true` is
`this.field1=void 0`
whereas with `allowDeclareFields: false`, no code is generated for that statement.

If we want to enable Metro users to enable `allowDeclareFields` for TypeScript and/or Flow, will that require adding a new metro-specific option, or can a user simply add a reference to the plugin in their own `babel.config.js`?